lib/db_connection/holder.ex

```diff
@@ -46,30 +65,106 @@ defmodule DBConnection.Holder do
     end
   end
 
-  @spec checkin(pool_ref :: any, state :: any) :: :ok
-  def checkin(pool_ref, state) do
+  @spec checkin(pool_ref :: any) :: :ok
+  def checkin(pool_ref) do
     now = System.monotonic_time(@time_unit)
-    done(pool_ref, :checkin, state, now)
+    done(pool_ref, :ok, :checkin, now)
   end
 
-  @spec disconnect(pool_ref :: any, err :: Exception.t(), state :: any) :: :ok
-  def disconnect(pool_ref, err, state) do
-    done(pool_ref, :disconnect, state, err)
+  @spec disconnect(pool_ref :: any, err :: Exception.t()) :: :ok
+  def disconnect(pool_ref, err) do
+    done(pool_ref, {:error, err}, :disconnect, err)
   end
 
-  @spec stop(pool_ref :: any, err :: Exception.t(), state :: any) :: :ok
-  def stop(pool_ref, err, state) do
-    done(pool_ref, :stop, state, err)
+  @spec stop(pool_ref :: any, err :: Exception.t()) :: :ok
+  def stop(pool_ref, err) do
+    done(pool_ref, {:error, err}, :stop, err)
   end
 
-  @spec update(pid, reference, module, term) :: t
-  def update(pool, ref, mod, state) do
-    holder = new(pool, ref, mod, state)
-    now = System.monotonic_time(@time_unit)
-    :ets.give_away(holder, pool, {:checkin, ref, now})
-    holder
+  @spec handle(pool_ref :: any, fun :: atom, args :: [term], Keyword.t) :: tuple
+  def handle(pool_ref, fun, args, opts) do
+    handle(:handle, pool_ref, fun, args, opts)
+  end
+
+  @spec cleanup(pool_ref :: any, fun :: atom, args :: [term], Keyword.t) :: tuple
+  def cleanup(pool_ref, fun, args, opts) do
+    handle(:cleanup, pool_ref, fun, args, opts)
+  end
+
+  defp handle(type, pool_ref, fun, args, opts) do
+    pool_ref(holder: holder) = pool_ref
+
+    try do
+      :ets.lookup(holder, :conn)
+    rescue
+      ArgumentError ->
+        msg = "connection is closed"
+        {:disconnect, DBConnection.ConnectionError.exception(msg), _state = :unused}
+    else
+      [conn(status: {:error, _})] ->
+        msg = "connection is closed"
+        {:disconnect, DBConnection.ConnectionError.exception(msg), _state = :unused}
+
+      [conn(status: :aborted)] when type != :cleanup ->
+        msg = "transaction rolling back"
+        {:disconnect, DBConnection.ConnectionError.exception(msg), _state = :unused}
+
+      [conn(module: module, state: state)] ->
+        try do
```

Comment:
Can we move this try to a separate function to removing the nesting?

Reply:
I will refactor this in the separate PR where we improve the errors.